# Software package and Image management
Notebook that details the different docker images, how to get them, what their purpose is and what software packages come default. Also gives examples on how to get new packages.
## Purpose and structure of images:
## Megatid
Based on the Debian Jessie (will likely move to alpine) linux image. Software packages installed are centered around taking sequence reads and processing them using any of the well established software packages (bowtie2, bwa, RSEM, Origami, HiCPro, etc) into commonly accepted output formats, such as FPKM, differential expression, normalized interactions, etc.
## Euplotid
Based on the Debian Jessie (moving to alpine) linux image. Includes machine learning PyTorch ready to go. It can train and test the Convolutional Neural Network (CNN) + Long Short Term Memory (LSTM).  Its loaded with the ability to call Insulated Neighborhoods using a louvian algorithm and then annotate each chromatin accessibility peak falling within them. The identities of each TF are estimated using the trained neural network. Eventually it outputs JSON files storing the annotated INs
## Minitid
Based on the minimal Alpine Linux image (starts at 4.5Mb!). On top are only the packages required in order to run plotly and jupyter in order to give a simple front end to the annotated insulated neighborhoods. Use this if you want to visualize the Insulated Neighborhood graphs.

## Get Docker
[**INSTALL DOCKER HERE**](https://www.docker.com/community-edition#/download")

The pipelines available and their capabilities are described in [Methods](docs/Methods.md)  which helps you pick the right of 3 Docker images.

Remember to define the correct directories when running the Docker image depending on your local OS machine, ex:

Whitehead (Linux):
~~~
    -v "/lab/solexa_public:/input_dir" \
    -v "/home/dborgesr/work_space/tmp:/tmp_dir" \
    -v "/home/dborgesr/work_space/out_dir:/output_dir" \
    -v "/home/dborgesr/work_space/annotation:/annotation_dir" \
~~~

## Pull and run your image!

* Megatid: process sequencing data into quantified values (FPKM,peaks,etc)
~~~ 
docker run --name megatid -p 8891:8891 -tid \
	-v "/your/input/directory:/input_dir" \
	-v "/your/temporary/directory/:/tmp_dir" \
	-v "/your/output/directory/:/output_dir" \
	-v "/your/annotation/directory/:/annotation_dir" \
	dborgesr/euplotid:megatid
~~~
* Euplotid: build and visualize INs and learn/predict TFs bound at CREs
~~~
docker run --name euplotid -p 8890:8890 -tid \
	-v "/your/input/directory:/input_dir" \
	-v "/your/temporary/directory/:/tmp_dir" \
	-v "/your/output/directory/:/output_dir" \
	-v "/your/annotation/directory/:/annotation_dir" \
	dborgesr/euplotid:euplotid
~~~
* Minitid: visualize and interact with built and annotated INs
~~~
docker run --name minitid -p 8892:8892 -tid \
	-v "/your/input/directory:/input_dir" \
	-v "/your/temporary/directory/:/tmp_dir" \
	-v "/your/output/directory/:/output_dir" \
	-v "/your/annotation/directory/:/annotation_dir" \
	dborgesr/euplotid:minitid
~~~
* Nanotid: ARM architecture image to build and visualize INs and learn/predict TFs bound at CREs
~~~
docker run --name nanotid -p 8893:8893 -tid \
	-v "/your/input/directory:/input_dir" \
	-v "/your/temporary/directory/:/tmp_dir" \
	-v "/your/output/directory/:/output_dir" \
	-v "/your/annotation/directory/:/annotation_dir" \
	dborgesr/euplotid:nanotid
~~~
Your Docker image should be running on your computer (local), now you can access and use it!

## Go to your image!
* Megatid <br> local:[http://localhost:8891](http://localhost:8891) Whitehead internal:[http://airstream:8891](http://airstream:8891)
* Euplotid <br> local:[http://localhost:8890](http://localhost:8890) Whitehead internal:[http://airstream:8890](http://airstream:8890)
* Minitid <br> local:[http://localhost:8892](http://localhost:8892) Whitehead internal:[http://airstream:8892](http://airstream:8892)
* Nanotid <br> local:[http://localhost:8893](http://localhost:8893) Whitehead internal:[http://airstream:8893](http://airstream:8893)

Each Docker image has different capabilities (packages installed in each Docker image are described in [packageManagement](docs/packageManagement.md))

## Pull images from dockerhub

In [2]:
!docker pull dborgesr/euplotid:megatid
!docker pull dborgesr/euplotid:euplotid
!docker pull dborgesr/euplotid:minitid

megatid: Pulling from dborgesr/euplotid

10a267c67f42: Pulling fs layer 
adf80874a519: Pulling fs layer 
4fc21761ddee: Pulling fs layer 
152e632d6b05: Pulling fs layer 
3c180d94b9e7: Pulling fs layer 
a17f00e5c574: Pulling fs layer 
879d1dd0a28a: Pulling fs layer 
ad4e14517bd2: Pulling fs layer 
76b27a6c70f5: Pulling fs layer 
76b27a6c70f5: Waiting fs layer 
2c568cb0bc51: Waiting fs layer 
08ad827b2363: Pulling fs layer 
08ad827b2363: Waiting fs layer 
262e6f42d76c: Waiting fs layer 
60d563535daf: Pulling fs layer 
60d563535daf: Waiting fs layer 
39ba9e3966c7: Pulling fs layer 
9cf6c92fbfe2: Pulling fs layer 
24d4b13d4479: Pull complete 69 kB/6.269 kBBDownloading 24.84 MB/157.9 MBDownloading  45.4 MB/850.6 MBadf80874a519: Downloading 71.88 MB/157.9 MBDownloading 396.3 MB/850.6 MBDownloading 49.73 MB/483.8 MBad4e14517bd2: Downloading 355.7 MB/1.031 GBDownloading 395.7 MB/1.031 GBDownloading 676.4 MB/850.6 MBDownloading 227.6 MB/483.8 MB152e632d6b05: Downloading 736.4 MB/850.6 MBDownload

## Run images

In [9]:
#Megatid
docker run --name megatid_cont -p 8891:8891 -tid \
    -v "/lab/solexa_public/:/input_dir" \
    -v "/home/dborgesr/work_space/tmp/:/tmp_dir" \
    -v "/home/dborgesr/work_space/out_dir/:/output_dir" \
    -v "/home/dborgesr/work_space/annotation:/annotation_dir" \
    dborgesr/euplotid:megatid
#Euplotid
docker run --name euplotid_cont -p 8890:8890 -tid \
    -v "/lab/solexa_public/:/input_dir" \
    -v "/home/dborgesr/work_space/tmp/:/tmp_dir" \
    -v "/home/dborgesr/work_space/out_dir/:/output_dir" \
    -v "/home/dborgesr/work_space/annotation:/annotation_dir" \
    dborgesr/euplotid:euplotid
#Minitid
docker run --name minitid_cont -p 8892:8892 -tid \
    -v "/lab/solexa_public/:/input_dir" \
    -v "/home/dborgesr/work_space/tmp/:/tmp_dir" \
    -v "/home/dborgesr/work_space/out_dir/:/output_dir" \
    -v "/home/dborgesr/work_space/annotation:/annotation_dir" \
    dborgesr/euplotid:minitid

ad128feaec2d7b05f9b6388389c87716dba5de8ca0d7706ca771070d26f1200c
9462f64d0dce2e54f90e8f5e16bd5750b646bb74e983fb369aa6631be66f433a


## Nuke Docker, wipe all containers and imagees, CAREFUL!

In [ ]:
!docker stop $(docker ps -a -q)
!docker rm $(docker ps -a -q)
!docker rmi $(docker images -a -q)

## Update Euplotid

In [1]:
!cd /root/Euplotid/ && git pull

remote: Counting objects: 87, done.
remote: Total 87 (delta 19), reused 19 (delta 19), pack-reused 68
Unpacking objects: 100% (87/87), done.
From https://github.com/dborgesr/Euplotid
   a31c928..26d34d8  master     -> origin/master
   9a755cd..6c46a04  gh-pages   -> origin/gh-pages
Updating a31c928..26d34d8
Fast-forward
 pipelines/Introduction.ipynb | 395 +++++++++++++++++++++++++++++++++++++++----
 1 file changed, 360 insertions(+), 35 deletions(-)


## Conda
Conda is an AWESOME software package handler and I use it to upkeep all these images, [here](https://conda.io/docs/_downloads/conda-cheatsheet.pdf) is a useful cheat sheet!
Conda can easily add many many packages and handles all the version handling and dirty backend. Also allows us to easily export our environment for full and fast replicability.

In [ ]:
#List all packages on current machine using conda
!conda list

In [10]:
#export settings which include all packages to a file which you can share to be loaded anywhere
!conda env export --name root > /tmp/cur_enviro.yaml

DEPRECATION: The default format will switch to columns in the future. You can use --format=(legacy|columns) (or define a format=(legacy|columns) in your pip.conf under the [list] section) to disable this warning.


In [ ]:
#Install new package and specify channel and version
!conda install -c bioconda samtools=1.4.1 

## Default packages in Docker Images


| Megatid                                | Euplotid                   | Minitid                  |
|----------------------------------------|----------------------------|--------------------------|
| _license=1.1                           | _license=1.1               | _nb_ext_conf=0.3.0       |
| _nb_ext_conf=0.3.0                     | _nb_ext_conf=0.3.0         | anaconda-client=1.6.3    |
| alabaster=0.7.10                       | alabaster=0.7.9            | asn1crypto=0.22.0        |
| anaconda=custom                        | anaconda=4.2.0             | bcftools=1.4.1           |
| anaconda-clean=1.1.0                   | anaconda-clean=1.0.0       | bedops=2.4.26            |
| anaconda-client=1.6.3                  | anaconda-client=1.5.1      | biopython=1.68           |
| anaconda-navigator=1.5.2               | anaconda-navigator=1.3.1   | bleach=1.5.0             |
| anaconda-project=0.6.0                 | argcomplete=1.0.0          | bzip2=1.0.6              |
| argcomplete=1.8.2                      | argparse=1.4.0             | cffi=1.10.0              |
| asn1crypto=0.22.0                      | astroid=1.4.7              | clyent=1.2.2             |
| astroid=1.4.8                          | astropy=1.2.1              | conda=4.3.18             |
| astropy=1.3.2                          | babel=2.3.4                | conda-env=2.6.0          |
| autoconf=2.69                          | backports=1.0              | cryptography=1.8.1       |
| babel=2.3.4                            | backports_abc=0.4          | curl=7.52.1              |
| backports=1.0                          | beautifulsoup4=4.5.1       | dbus=1.10.10             |
| bamtools=2.4.0                         | biopython=1.69             | decorator=4.0.11         |
| bash_kernel=0.6                        | bitarray=0.8.1             | entrypoints=0.2.2        |
| beautifulsoup4=4.5.3                   | blaze=0.10.1               | expat=2.1.0              |
| bedops=2.4.26                          | bokeh=0.12.2               | fontconfig=2.12.1        |
| bedtools=2.26.0                        | boto=2.42.0                | freetype=2.5.5           |
| bitarray=0.8.1                         | bottleneck=1.1.0           | git=2.11.1               |
| blas=1.1                               | bx-python=0.7.3            | glib=2.50.2              |
| blat=35                                | cairo=1.12.18              | gst-plugins-base=1.8.0   |
| blaze=0.10.1                           | cdecimal=2.3               | gstreamer=1.8.0          |
| bleach=1.4.2                           | cffi=1.7.0                 | html5lib=0.999           |
| bokeh=0.12.5                           | chest=0.2.3                | htslib=1.4.1             |
| boto=2.46.1                            | click=6.6                  | icu=54.1                 |
| boto3=1.4.4                            | cloudpickle=0.2.1          | idna=2.5                 |
| botocore=1.5.39                        | clyent=1.2.2               | ipykernel=4.6.1          |
| bottleneck=1.2.0                       | colorama=0.3.7             | ipython=6.0.0            |
| bowtie=1.2.0                           | conda=4.2.9                | ipython_genutils=0.2.0   |
| bowtie2=2.3.2                          | conda-build=2.0.2          | ipywidgets=6.0.0         |
| bwa=0.7.15                             | configobj=5.0.6            | jbig=2.1                 |
| bzip2=1.0.6                            | configparser=3.5.0         | jedi=0.10.2              |
| ca-certificates=2017.4.17              | contextlib2=0.5.3          | jinja2=2.9.6             |
| cairo=1.14.6                           | cryptography=1.5           | joblib=0.11              |
| certifi=2017.4.17                      | curl=7.49.0                | jpeg=9b                  |
| cffi=1.10.0                            | cycler=0.10.0              | jsonschema=2.6.0         |
| chardet=3.0.2                          | cython=0.24.1              | jupyter=1.0.0            |
| chest=0.2.3                            | cytoolz=0.8.0              | jupyter_client=5.0.1     |
| click=6.7                              | dask=0.11.0                | jupyter_console=5.1.0    |
| cloog=0.18.0                           | datashape=0.5.2            | jupyter_core=4.3.0       |
| cloudpickle=0.2.2                      | dbus=1.10.10               | libffi=3.2.1             |
| clyent=1.2.1                           | decorator=4.0.10           | libgcc=5.2.0             |
| colorama=0.3.9                         | dill=0.2.5                 | libgfortran=3.0.0        |
| conda=4.2.13                           | docutils=0.12              | libiconv=1.14            |
| conda-build=2.1.10                     | dynd-python=0.7.2          | libpng=1.6.27            |
| conda-env=2.6.0                        | entrypoints=0.2.2          | libsodium=1.0.10         |
| conda-verify=2.0.0                     | enum34=1.1.6               | libtiff=4.0.6            |
| configobj=5.0.6                        | et_xmlfile=1.0.1           | libxcb=1.12              |
| contextlib2=0.5.5                      | expat=2.1.0                | libxml2=2.9.4            |
| cryptography=1.8.1                     | fastcache=1.0.2            | markupsafe=0.23          |
| curl=7.52.1                            | filelock=2.0.6             | mistune=0.7.4            |
| cycler=0.10.0                          | flask=0.11.1               | mkl=2017.0.1             |
| cython=0.25.2                          | flask-cors=2.1.2           | mmtf-python=1.0.5        |
| cytoolz=0.8.2                          | fontconfig=2.11.1          | msgpack-python=0.4.8     |
| dask=0.14.3                            | freetype=2.5.5             | mygene=2.3.0             |
| datashape=0.5.4                        | funcsigs=1.0.2             | myvariant=0.3.1          |
| dbus=1.10.10                           | functools32=3.2.3.2        | nb_anacondacloud=1.2.0   |
| decorator=4.0.11                       | futures=3.0.5              | nb_conda=2.0.0           |
| dill=0.2.6                             | get_terminal_size=1.0.0    | nb_conda_kernels=2.0.0   |
| distributed=1.16.3                     | gevent=1.1.2               | nbconvert=5.1.1          |
| docutils=0.13.1                        | Ghost.py=0.2.3             | nbformat=4.3.0           |
| dynd-python=0.7.2                      | glib=2.43.0                | nbpresent=3.0.2          |
| entrypoints=0.2.2                      | greenlet=0.4.10            | networkx=1.11            |
| et_xmlfile=1.0.1                       | grin=1.2.1                 | notebook=5.0.0           |
| expat=2.1.0                            | gst-plugins-base=1.8.0     | numpy=1.12.1             |
| fastcache=1.0.2                        | gstreamer=1.8.0            | olefile=0.44             |
| fastqc=0.11.5                          | h5py=2.6.0                 | openssl=1.0.2k           |
| filelock=2.0.6                         | harfbuzz=0.9.39            | packaging=16.8           |
| flask=0.12                             | hdf5=1.8.17                | pandas=0.20.1            |
| flask-cors=3.0.2                       | heapdict=1.0.0             | pandocfilters=1.4.1      |
| fontconfig=2.12.1                      | html5lib=0.999999999       | path.py=10.3.1           |
| fortran-magic=0.7                      | icu=54.1                   | pcre=8.39                |
| freetype=2.7                           | idna=2.1                   | pexpect=4.2.1            |
| gcc=4.8.5                              | imagesize=0.7.1            | pickleshare=0.7.4        |
| get_terminal_size=1.0.0                | ipaddress=1.0.16           | pillow=4.1.1             |
| gettext=0.19.7                         | ipykernel=4.5.0            | pip=9.0.1                |
| gevent=1.1rc4                          | ipython=5.1.0              | plotly=1.12.9            |
| ghostscript=9.18                       | ipython_genutils=0.1.0     | prompt_toolkit=1.0.14    |
| giflib=5.1.4                           | ipywidgets=5.2.2           | ptyprocess=0.5.1         |
| git=2.12.2                             | itsdangerous=0.24          | pycosat=0.6.2            |
| glib=2.51.4                            | jbig=2.1                   | pycparser=2.17           |
| gmp=6.1.2                              | jdcal=1.2                  | pycrypto=2.6.1           |
| graphite2=1.3.9                        | jedi=0.9.0                 | pygments=2.2.0           |
| graphviz=2.38.0                        | jinja2=2.8                 | pyliftover=0.3           |
| greenlet=0.4.12                        | joblib=0.11                | pyopenssl=17.0.0         |
| gsl=2.2.1                              | jpeg=8d                    | pyparsing=2.1.4          |
| gst-plugins-base=1.8.0                 | jsonschema=2.5.1           | pyqt=5.6.0               |
| gstreamer=1.8.0                        | jupyter=1.0.0              | pysam=0.11.2.1           |
| h5py=2.7.0                             | jupyter_client=4.4.0       | python=3.5.3             |
| harfbuzz=1.4.3                         | jupyter_console=5.0.0      | python-dateutil=2.6.0    |
| hdf5=1.8.17                            | jupyter_core=4.2.0         | pytorch=0.1.12           |
| heapdict=1.0.0                         | lazy-object-proxy=1.2.1    | pytz=2017.2              |
| homer=4.8.3                            | libdynd=0.7.2              | pyyaml=3.12              |
| html5lib=0.999999999                   | libffi=3.2.1               | pyzmq=16.0.2             |
| htslib=1.4.1                           | libgcc=4.8.5               | qt=5.6.2                 |
| icu=58.1                               | libgfortran=3.0.0          | qtconsole=4.3.0          |
| idna=2.5                               | libpng=1.6.22              | readline=6.2             |
| imageio=2.1.2                          | libsodium=1.0.10           | reportlab=3.4.0          |
| imagesize=0.7.1                        | libtiff=4.0.6              | requests=2.14.2          |
| ipykernel=4.6.1                        | libxcb=1.12                | ruamel_yaml=0.11.14      |
| ipyparallel=6.0.2                      | libxml2=2.9.2              | samtools=1.4.1           |
| ipython=6.0.0                          | libxslt=1.1.28             | scipy=0.19.0             |
| ipython_genutils=0.2.0                 | llvmlite=0.13.0            | setuptools=27.2.0        |
| ipywidgets=6.0.0                       | locket=0.2.0               | simplegeneric=0.8.1      |
| isl=0.12.2                             | lxml=3.6.4                 | sip=4.18                 |
| isort=4.2.5                            | MACS2=2.1.1.20160309       | six=1.10.0               |
| itsdangerous=0.24                      | markupsafe=0.23            | sqlite=3.13.0            |
| java-jdk=8.0.92                        | matplotlib=1.5.3           | terminado=0.6            |
| jbig=2.1                               | mechanize=0.3.3            | testpath=0.3             |
| jdcal=1.3                              | mistune=0.7.3              | tk=8.5.18                |
| jedi=0.10.0                            | mkl=11.3.3                 | torchvision=0.1.8        |
| jinja2=2.9.5                           | mkl-service=1.1.2          | tornado=4.5.1            |
| jmespath=0.9.0                         | mpmath=0.19                | traitlets=4.3.2          |
| jpeg=9b                                | multipledispatch=0.4.8     | wcwidth=0.1.7            |
| jsonschema=2.5.1                       | mygene=3.0.0               | wheel=0.29.0             |
| jupyter=1.0.0                          | myvariant=0.3.1            | widgetsnbextension=2.0.0 |
| jupyter-c-kernel=1.0.0                 | nb_anacondacloud=1.2.0     | xz=5.2.2                 |
| jupyter_client=5.0.1                   | nb_conda=2.0.0             | yaml=0.1.6               |
| jupyter_console=5.1.0                  | nb_conda_kernels=2.0.0     | zeromq=4.1.5             |
| jupyter_core=4.3.0                     | nbbrowserpdf=0.2.0         | zlib=1.2.8               |
| lazy-object-proxy=1.3.0                | nbconvert=4.2.0            |                          |
| libdynd=0.7.2                          | nbformat=4.1.0             |                          |
| libffi=3.2.1                           | nbformat=4.3.0             |                          |
| libgcc=5.2.0                           | nbpresent=3.0.2            |                          |
| libgd=2.2.4                            | networkx=1.11              |                          |
| libgfortran=3.0.0                      | nltk=3.2.1                 |                          |
| libiconv=1.14                          | nose=1.3.7                 |                          |
| libmagic=5.25                          | notebook=4.2.3             |                          |
| libpng=1.6.28                          | numba=0.28.1               |                          |
| libsodium=1.0.10                       | numexpr=2.6.1              |                          |
| libtiff=4.0.6                          | numpy=1.11.1               |                          |
| libwebp=0.5.2                          | odo=0.5.0                  |                          |
| libxcb=1.12                            | openpyxl=2.3.2             |                          |
| libxml2=2.9.4                          | openssl=1.0.2j             |                          |
| libxslt=1.1.29                         | pandas=0.18.1              |                          |
| llvmlite=0.18.0                        | partd=0.3.6                |                          |
| locket=0.2.0                           | patchelf=0.9               |                          |
| lxml=3.7.3                             | path.py=8.2.1              |                          |
| lzo=2.06                               | pathlib2=2.1.0             |                          |
| m4=1.4.17                              | patsy=0.4.1                |                          |
| make=4.2.1                             | pdfrw=0.3                  |                          |
| markupsafe=0.23                        | pep8=1.7.0                 |                          |
| matplotlib=2.0.2                       | pexpect=4.0.1              |                          |
| mccabe=0.6.1                           | pickleshare=0.7.4          |                          |
| mistune=0.7.4                          | pillow=3.3.1               |                          |
| mkl=2017.0.1                           | pip=9.0.1                  |                          |
| mkl-service=1.1.2                      | pip=8.1.2                  |                          |
| mpc=1.0.3                              | pixman=0.32.6              |                          |
| mpfr=3.1.5                             | pkginfo=1.3.2              |                          |
| mpmath=0.19                            | plotly=2.0.8               |                          |
| msgpack-python=0.4.8                   | ply=3.9                    |                          |
| multipledispatch=0.4.9                 | prompt_toolkit=1.0.3       |                          |
| nb_anacondacloud=1.4.0                 | psutil=4.3.1               |                          |
| nb_conda=2.2.0                         | ptyprocess=0.5.1           |                          |
| nb_conda_kernels=2.1.0                 | py=1.4.31                  |                          |
| nbconvert=5.1.1                        | pyasn1=0.1.9               |                          |
| nbformat=4.3.0                         | pycairo=1.10.0             |                          |
| nbpresent=3.0.2                        | pycosat=0.6.1              |                          |
| ncurses=5.9                            | pycparser=2.14             |                          |
| networkx=1.11                          | pycrypto=2.6.1             |                          |
| nltk=3.2.2                             | pycurl=7.43.0              |                          |
| nose=1.3.7                             | pyflakes=1.3.0             |                          |
| notebook=5.0.0                         | pygments=2.1.3             |                          |
| numba=0.33.0                           | pyliftover=0.3             |                          |
| numexpr=2.6.2                          | pylint=1.5.4               |                          |
| numpy=1.12.1                           | pyopenssl=16.0.0           |                          |
| numpydoc=0.6.0                         | pyparsing=2.1.4            |                          |
| odo=0.5.0                              | PyPDF2=1.26.0              |                          |
| olefile=0.44                           | pyqt=5.6.0                 |                          |
| openblas=0.2.19                        | pysam=0.11.2.1             |                          |
| openjdk=8.0.121                        | pytables=3.2.3.1           |                          |
| openpyxl=2.4.0                         | pytest=2.9.2               |                          |
| openssl=1.0.2k                         | python=2.7.12              |                          |
| packaging=16.8                         | python-dateutil=2.5.3      |                          |
| pandas=0.20.1                          | python-louvain=0.6         |                          |
| pandoc=1.19.2                          | pytz=2016.6.1              |                          |
| pandocfilters=1.4.1                    | pyyaml=3.12                |                          |
| pango=1.40.4                           | pyzmq=15.4.0               |                          |
| partd=0.3.8                            | qt=5.6.0                   |                          |
| patchelf=0.9                           | qtawesome=0.3.3            |                          |
| path.py=8.2.1                          | qtconsole=4.2.1            |                          |
| pathlib2=2.2.1                         | qtpy=1.1.2                 |                          |
| patsy=0.4.1                            | readline=6.2               |                          |
| pcre=8.39                              | redis=3.2.0                |                          |
| pep8=1.7.0                             | redis-py=2.10.5            |                          |
| perl=5.22.0.1                          | reportlab=3.4.0            |                          |
| perl-aceperl=1.92                      | requests=2.11.1            |                          |
| perl-algorithm-munkres=0.08            | rope=0.9.4                 |                          |
| perl-app-cpanminus=1.7039              | rst2pdf-0.93.dev=r0        |                          |
| perl-appconfig=1.71                    | ruamel_yaml=0.11.14        |                          |
| perl-array-compare=2.11                | scikit-image=0.12.3        |                          |
| perl-bio-asn1-entrezgene=1.72          | scikit-learn=0.17.1        |                          |
| perl-bio-featureio=1.6.905             | scipy=0.18.1               |                          |
| perl-bio-phylo=0.58                    | seaborn=0.7.1              |                          |
| perl-bio-samtools=1.43                 | setuptools=27.2.0          |                          |
| perl-bioperl=1.6.924                   | simplegeneric=0.8.1        |                          |
| perl-bioperl-core=1.6.924              | singledispatch=3.4.0.3     |                          |
| perl-bioperl-run=1.006900              | sip=4.18                   |                          |
| perl-cgi=4.22                          | six=1.10.0                 |                          |
| perl-class-inspector=1.28              | snowballstemmer=1.2.1      |                          |
| perl-class-method-modifiers=2.11       | sockjs-tornado=1.0.3       |                          |
| perl-clone=0.38                        | sphinx=1.4.6               |                          |
| perl-common-sense=3.74                 | spyder=3.0.0               |                          |
| perl-convert-binary-c=0.78             | sqlalchemy=1.0.13          |                          |
| perl-convert-binhex=1.125              | sqlite=3.13.0              |                          |
| perl-crypt-rc4=2.02                    | ssl_match_hostname=3.4.0.2 |                          |
| perl-data-stag=0.14                    | statsmodels=0.6.1          |                          |
| perl-dbd-sqlite=1.50                   | sympy=1.0                  |                          |
| perl-dbi=1.636                         | terminado=0.6              |                          |
| perl-devel-globaldestruction=0.13      | tk=8.5.18                  |                          |
| perl-digest-hmac=1.03                  | toolz=0.8.0                |                          |
| perl-digest-perl-md5=1.9               | tornado=4.4.1              |                          |
| perl-email-date-format=1.005           | traitlets=4.3.0            |                          |
| perl-encode-locale=1.05                | unicodecsv=0.14.1          |                          |
| perl-error=0.17024                     | uuid=1.30                  |                          |
| perl-exporter-tiny=0.042               | Wand=0.4.4                 |                          |
| perl-file-listing=6.04                 | wcwidth=0.1.7              |                          |
| perl-file-slurp-tiny=0.004             | webencodings=0.5.1         |                          |
| perl-file-sort=1.01                    | weblogo=3.5.0              |                          |
| perl-file-which=1.20                   | werkzeug=0.11.11           |                          |
| perl-font-afm=1.20                     | wheel=0.29.0               |                          |
| perl-gd=2.56                           | widgetsnbextension=1.2.6   |                          |
| perl-graph=0.9704                      | wrapt=1.10.6               |                          |
| perl-graphviz=2.20                     | xlrd=1.0.0                 |                          |
| perl-html-element-extended=1.18        | xlsxwriter=0.9.3           |                          |
| perl-html-entities-numbered=0.04       | xlwt=1.1.2                 |                          |
| perl-html-formatter=2.14               | xz=5.2.2                   |                          |
| perl-html-parser=3.72                  | yaml=0.1.6                 |                          |
| perl-html-tableextract=2.13            | zeromq=4.1.4               |                          |
| perl-html-tagset=3.20                  | zlib=1.2.8                 |                          |
| perl-html-tidy=1.56                    |                            |                          |
| perl-html-tree=5.03                    |                            |                          |
| perl-html-treebuilder-xpath=0.14       |                            |                          |
| perl-http-cookies=6.01                 |                            |                          |
| perl-http-daemon=6.01                  |                            |                          |
| perl-http-date=6.02                    |                            |                          |
| perl-http-message=6.11                 |                            |                          |
| perl-http-negotiate=6.01               |                            |                          |
| perl-image-info=1.38                   |                            |                          |
| perl-image-size=3.300                  |                            |                          |
| perl-io-html=1.001                     |                            |                          |
| perl-io-sessiondata=1.03               |                            |                          |
| perl-io-socket-ssl=2.024               |                            |                          |
| perl-io-string=1.08                    |                            |                          |
| perl-io-stringy=2.111                  |                            |                          |
| perl-io-tty=1.12                       |                            |                          |
| perl-ipc-run=0.94                      |                            |                          |
| perl-jcode=2.07                        |                            |                          |
| perl-json=2.90                         |                            |                          |
| perl-json-xs=2.34                      |                            |                          |
| perl-libwww-perl=6.15                  |                            |                          |
| perl-libxml-perl=0.08                  |                            |                          |
| perl-list-moreutils=0.413              |                            |                          |
| perl-lwp-mediatypes=6.02               |                            |                          |
| perl-lwp-protocol-https=6.06           |                            |                          |
| perl-lwp-simple=6.15                   |                            |                          |
| perl-mailtools=2.14                    |                            |                          |
| perl-math-cdf=0.1                      |                            |                          |
| perl-math-derivative=0.04              |                            |                          |
| perl-math-random=0.72                  |                            |                          |
| perl-math-spline=0.02                  |                            |                          |
| perl-mime-lite=3.030                   |                            |                          |
| perl-mime-tools=5.507                  |                            |                          |
| perl-mime-types=2.12                   |                            |                          |
| perl-mldbm=2.05                        |                            |                          |
| perl-module-runtime=0.014              |                            |                          |
| perl-moo=2.001000                      |                            |                          |
| perl-net-http=6.09                     |                            |                          |
| perl-net-ssleay=1.74                   |                            |                          |
| perl-ntlm=1.09                         |                            |                          |
| perl-ole-storage_lite=0.19             |                            |                          |
| perl-parse-recdescent=1.967013         |                            |                          |
| perl-pdf-api2=2.025                    |                            |                          |
| perl-postscript=0.06                   |                            |                          |
| perl-role-tiny=2.000001                |                            |                          |
| perl-set-scalar=1.29                   |                            |                          |
| perl-soap-lite=1.19                    |                            |                          |
| perl-sort-naturally=1.03               |                            |                          |
| perl-spreadsheet-parseexcel=0.65       |                            |                          |
| perl-spreadsheet-writeexcel=2.40       |                            |                          |
| perl-statistics-descriptive=3.0612     |                            |                          |
| perl-sub-exporter-progressive=0.001011 |                            |                          |
| perl-svg=2.64                          |                            |                          |
| perl-svg-graph=0.02                    |                            |                          |
| perl-task-weaken=1.04                  |                            |                          |
| perl-template-toolkit=2.26             |                            |                          |
| perl-test-leaktrace=0.15               |                            |                          |
| perl-test-pod=1.51                     |                            |                          |
| perl-threaded=5.22.0                   |                            |                          |
| perl-tie-ixhash=1.23                   |                            |                          |
| perl-timedate=2.30                     |                            |                          |
| perl-tree-dag_node=1.29                |                            |                          |
| perl-type-tiny=1.000005                |                            |                          |
| perl-unicode-map=0.112                 |                            |                          |
| perl-uri=1.71                          |                            |                          |
| perl-www-robotrules=6.02               |                            |                          |
| perl-xml-dom=1.45                      |                            |                          |
| perl-xml-dom-xpath=0.14                |                            |                          |
| perl-xml-filter-buffertext=1.01        |                            |                          |
| perl-xml-libxml=2.0124                 |                            |                          |
| perl-xml-libxslt=1.94                  |                            |                          |
| perl-xml-namespacesupport=1.11         |                            |                          |
| perl-xml-parser=2.44                   |                            |                          |
| perl-xml-regexp=0.04                   |                            |                          |
| perl-xml-sax=0.99                      |                            |                          |
| perl-xml-sax-base=1.08                 |                            |                          |
| perl-xml-sax-expat=0.51                |                            |                          |
| perl-xml-sax-writer=0.56               |                            |                          |
| perl-xml-simple=2.22                   |                            |                          |
| perl-xml-twig=3.49                     |                            |                          |
| perl-xml-writer=0.625                  |                            |                          |
| perl-xml-xpath=1.33                    |                            |                          |
| perl-xml-xpathengine=0.14              |                            |                          |
| perl-yaml=1.18                         |                            |                          |
| pexpect=4.2.1                          |                            |                          |
| pickleshare=0.7.3                      |                            |                          |
| pillow=4.1.1                           |                            |                          |
| pip=9.0.1                              |                            |                          |
| pixman=0.34.0                          |                            |                          |
| pkginfo=1.2.1                          |                            |                          |
| plotly=2.0.7                           |                            |                          |
| ply=3.9                                |                            |                          |
| prompt_toolkit=1.0.14                  |                            |                          |
| psutil=5.2.1                           |                            |                          |
| ptyprocess=0.5.1                       |                            |                          |
| py=1.4.33                              |                            |                          |
| pyasn1=0.2.3                           |                            |                          |
| pycosat=0.6.1                          |                            |                          |
| pycparser=2.17                         |                            |                          |
| pycrypto=2.6.1                         |                            |                          |
| pycurl=7.43.0                          |                            |                          |
| pyflakes=1.5.0                         |                            |                          |
| pygments=2.2.0                         |                            |                          |
| pylint=1.6.5                           |                            |                          |
| pyopenssl=16.2.0                       |                            |                          |
| pyparsing=2.2.0                        |                            |                          |
| pyqt=5.6.0                             |                            |                          |
| pytables=3.4.2                         |                            |                          |
| pytest=3.0.7                           |                            |                          |
| python=3.5.3                           |                            |                          |
| python-dateutil=2.6.0                  |                            |                          |
| pytz=2017.2                            |                            |                          |
| pywavelets=0.5.2                       |                            |                          |
| pyyaml=3.12                            |                            |                          |
| pyzmq=16.0.2                           |                            |                          |
| qt=5.6.2                               |                            |                          |
| qtawesome=0.4.4                        |                            |                          |
| qtconsole=4.3.0                        |                            |                          |
| qtpy=1.2.1                             |                            |                          |
| r-assertthat=0.1                       |                            |                          |
| r-backports=1.0.4                      |                            |                          |
| r-base=3.3.2                           |                            |                          |
| r-base64enc=0.1_3                      |                            |                          |
| r-bh=1.62.0_1                          |                            |                          |
| r-bitops=1.0_6                         |                            |                          |
| r-boot=1.3_18                          |                            |                          |
| r-broom=0.4.1                          |                            |                          |
| r-car=2.1_4                            |                            |                          |
| r-caret=6.0_73                         |                            |                          |
| r-catools=1.17.1                       |                            |                          |
| r-class=7.3_14                         |                            |                          |
| r-cluster=2.0.5                        |                            |                          |
| r-codetools=0.2_15                     |                            |                          |
| r-colorspace=1.3_1                     |                            |                          |
| r-crayon=1.3.2                         |                            |                          |
| r-curl=2.3                             |                            |                          |
| r-data.table=1.10.0                    |                            |                          |
| r-dbi=0.5_1                            |                            |                          |
| r-dichromat=2.0_0                      |                            |                          |
| r-digest=0.6.10                        |                            |                          |
| r-dplyr=0.5.0                          |                            |                          |
| r-essentials=1.5.2                     |                            |                          |
| r-evaluate=0.10                        |                            |                          |
| r-forcats=0.1.1                        |                            |                          |
| r-foreach=1.4.3                        |                            |                          |
| r-foreign=0.8_67                       |                            |                          |
| r-formatr=1.4                          |                            |                          |
| r-ggplot2=2.2.0                        |                            |                          |
| r-gistr=0.3.6                          |                            |                          |
| r-glmnet=2.0_5                         |                            |                          |
| r-gtable=0.2.0                         |                            |                          |
| r-haven=1.0.0                          |                            |                          |
| r-hexbin=1.27.1                        |                            |                          |
| r-highr=0.6                            |                            |                          |
| r-hms=0.3                              |                            |                          |
| r-htmltools=0.3.5                      |                            |                          |
| r-htmlwidgets=0.8                      |                            |                          |
| r-httpuv=1.3.3                         |                            |                          |
| r-httr=1.2.1                           |                            |                          |
| r-irdisplay=0.4.4                      |                            |                          |
| r-irkernel=0.7.1                       |                            |                          |
| r-iterators=1.0.8                      |                            |                          |
| r-jsonlite=1.1                         |                            |                          |
| r-kernsmooth=2.23_15                   |                            |                          |
| r-knitr=1.15.1                         |                            |                          |
| r-labeling=0.3                         |                            |                          |
| r-lattice=0.20_34                      |                            |                          |
| r-lazyeval=0.2.0                       |                            |                          |
| r-lme4=1.1_12                          |                            |                          |
| r-lubridate=1.6.0                      |                            |                          |
| r-magrittr=1.5                         |                            |                          |
| r-maps=3.1.1                           |                            |                          |
| r-markdown=0.7.7                       |                            |                          |
| r-mass=7.3_45                          |                            |                          |
| r-matrix=1.2_7.1                       |                            |                          |
| r-matrixmodels=0.4_1                   |                            |                          |
| r-mgcv=1.8_16                          |                            |                          |
| r-mime=0.5                             |                            |                          |
| r-minqa=1.2.4                          |                            |                          |
| r-mnormt=1.5_5                         |                            |                          |
| r-modelmetrics=1.1.0                   |                            |                          |
| r-modelr=0.1.0                         |                            |                          |
| r-munsell=0.4.3                        |                            |                          |
| r-nlme=3.1_128                         |                            |                          |
| r-nloptr=1.0.4                         |                            |                          |
| r-nnet=7.3_12                          |                            |                          |
| r-openssl=0.9.5                        |                            |                          |
| r-pbdzmq=0.2_4                         |                            |                          |
| r-pbkrtest=0.4_6                       |                            |                          |
| r-plyr=1.8.4                           |                            |                          |
| r-pryr=0.1.2                           |                            |                          |
| r-psych=1.6.9                          |                            |                          |
| r-purrr=0.2.2                          |                            |                          |
| r-quantmod=0.4_7                       |                            |                          |
| r-quantreg=5.29                        |                            |                          |
| r-r6=2.2.0                             |                            |                          |
| r-randomforest=4.6_12                  |                            |                          |
| r-rbokeh=0.5.0                         |                            |                          |
| r-rcolorbrewer=1.1_2                   |                            |                          |
| r-rcpp=0.12.8                          |                            |                          |
| r-rcppeigen=0.3.2.9.0                  |                            |                          |
| r-readr=1.0.0                          |                            |                          |
| r-readxl=0.1.1                         |                            |                          |
| r-recommended=3.3.2                    |                            |                          |
| r-repr=0.10                            |                            |                          |
| r-reshape2=1.4.2                       |                            |                          |
| r-rmarkdown=1.3                        |                            |                          |
| r-rpart=4.1_10                         |                            |                          |
| r-rprojroot=1.1                        |                            |                          |
| r-rvest=0.3.2                          |                            |                          |
| r-scales=0.4.1                         |                            |                          |
| r-selectr=0.3_0                        |                            |                          |
| r-shiny=0.14.2                         |                            |                          |
| r-sourcetools=0.1.5                    |                            |                          |
| r-sparsem=1.74                         |                            |                          |
| r-spatial=7.3_11                       |                            |                          |
| r-stringi=1.1.2                        |                            |                          |
| r-stringr=1.1.0                        |                            |                          |
| r-survival=2.40_1                      |                            |                          |
| r-tibble=1.2                           |                            |                          |
| r-tidyr=0.6.0                          |                            |                          |
| r-tidyverse=1.0.0                      |                            |                          |
| r-ttr=0.23_1                           |                            |                          |
| r-uuid=0.1_2                           |                            |                          |
| r-xml2=1.0.0                           |                            |                          |
| r-xtable=1.8_2                         |                            |                          |
| r-xts=0.9_7                            |                            |                          |
| r-yaml=2.1.14                          |                            |                          |
| r-zoo=1.7_13                           |                            |                          |
| readline=6.2                           |                            |                          |
| redis=3.2.0                            |                            |                          |
| redis-py=2.10.5                        |                            |                          |
| requests=2.13.0                        |                            |                          |
| rope=0.10.5                            |                            |                          |
| ruamel_yaml=0.11.14                    |                            |                          |
| s3fs=0.1.0                             |                            |                          |
| s3transfer=0.1.10                      |                            |                          |
| samtools=1.4.1                         |                            |                          |
| scikit-image=0.13.0                    |                            |                          |
| scikit-learn=0.18.1                    |                            |                          |
| scipy=0.19.0                           |                            |                          |
| seaborn=0.7.1                          |                            |                          |
| setuptools=33.1.1                      |                            |                          |
| sickle=0.5                             |                            |                          |
| simplegeneric=0.8.1                    |                            |                          |
| singledispatch=3.4.0.3                 |                            |                          |
| sip=4.18                               |                            |                          |
| six=1.10.0                             |                            |                          |
| snowballstemmer=1.2.1                  |                            |                          |
| sockjs-tornado=1.0.3                   |                            |                          |
| sortedcontainers=1.5.3                 |                            |                          |
| sphinx=1.6.1                           |                            |                          |
| sphinxcontrib-websupport=1.0.1         |                            |                          |
| spyder=3.1.4                           |                            |                          |
| sqlalchemy=1.1.5                       |                            |                          |
| sqlite=3.13.0                          |                            |                          |
| sra-tools=2.8.1                        |                            |                          |
| star=2.5.3a                            |                            |                          |
| statsmodels=0.8.0                      |                            |                          |
| sympy=1.0                              |                            |                          |
| tbb=4.4_20150728                       |                            |                          |
| tblib=1.3.2                            |                            |                          |
| terminado=0.6                          |                            |                          |
| testpath=0.3                           |                            |                          |
| tidyp=1.04                             |                            |                          |
| tk=8.5.19                              |                            |                          |
| toolz=0.8.2                            |                            |                          |
| tornado=4.5.1                          |                            |                          |
| traitlets=4.3.2                        |                            |                          |
| trimmomatic=0.36                       |                            |                          |
| typing=3.5.2.2                         |                            |                          |
| unicodecsv=0.14.1                      |                            |                          |
| wcwidth=0.1.7                          |                            |                          |
| webencodings=0.5                       |                            |                          |
| weblogo=2.8.2                          |                            |                          |
| werkzeug=0.11.10                       |                            |                          |
| wheel=0.29.0                           |                            |                          |
| widgetsnbextension=2.0.0               |                            |                          |
| wkhtmltopdf=0.12.3                     |                            |                          |
| wrapt=1.10.8                           |                            |                          |
| xlrd=1.0.0                             |                            |                          |
| xlsxwriter=0.9.6                       |                            |                          |
| xlwt=1.2.0                             |                            |                          |
| xorg-libxau=1.0.8                      |                            |                          |
| xorg-libxdmcp=1.1.2                    |                            |                          |
| xz=5.2.2                               |                            |                          |
| yaml=0.1.6                             |                            |                          |
| zeromq=4.1.5                           |                            |                          |
| zict=0.1.2                             |                            |                          |
| zlib=1.2.8                             |                            |                          |